public/lessons/db/mongoose.md

```diff
@@ -0,0 +1,106 @@
+# Overview
+When writing to a database, it is often important to ensure the data types/fields are expected.
+
+Enter schema-validation with **Mongoose**. 
+
+[Mongoose](https://mongoosejs.com/) is an object data modeling (ODM) library.
+ 
+Or in their words,
+
+        "A straight-forward, schema-based solution to model your application data"
+
+# Getting Started
+* Mongoose replaces the `mongodb` driver
+* Runs `mongodb` queries under the hood
+
+Make a new directory named `mongoose-example` 
+
+```
+cd mongoose-example
+npm init -y
+npm install mongoose
+```
+# Getting Started, cont.
+Create a file called `mongoose.js`
+In it, write the following code.
+
+```javascript
+const mongoose = require('mongoose')
+/* 
+pass { useNewUrlParser: true, useUnifiedTopology: true } as a second 
+argument to the connect method to avoid deprecation warnings
+*/
+mongoose.connect('mongodb://localhost:27017/example_db') 
+const db = mongoose.connection
+
+db.on('error', console.error.bind(console, 'connection error:'))
+```
+
+# Concept: Schemas
+A database schema outlines the expected structure of the data that will be inserted into a *collection*
+
+A database schema can also define *methods* on the documents being inserted. 
+
+* *defines* what that data should look like, and how it should behave.
+* *enforces* that definition.
+
+```javascript
+const studentSchema = new mongoose.Schema({
+    name: String,
+    age: Number,
+    hobbies: Array,
+    current: Boolean
+})
+```
+The code above defines a simple schema that expects a certain datatype for the given field.
+A comprehensive list can be found [here](https://mongoosejs.com/docs/guide.html#definition).
+
+# Concept: Model
+While the definition of the data's structure is held in the Schema, a *Model* actually handles the work.
+
+Take our previous schema, `studentSchema`. Let's create a *model* from that schema, and call it `Student`.
+
+```javascript
+const Student = mongoose.model('Student', studentSchema)
+```
+Think of `Student` as a *Class* with enforcement run by the *Schema* underneath.
+
+We can create an *instance* of the `Student` model like we would any class!
+
+```javascript
+const paul = new Student({ name: 'Paul', age: 29, hobbies: ['guitar', 'd&d', 'coding'] })
+// call the save() method on a model instance (document) to insert it to the collection 
+paul.save()
+```
+Or a version with error handling
+
+```javascript
+ paul.save((err, paul) => {
+        if (err) return console.error(err);
+        console.log("document inserted!")
+    })
```

Comment:
One of the comments from the last cohort was that we should have less shorthand in our code examples, so just expand this into slightly more verbose `if...else` syntax.